apps/server/src/state.ts

```diff
@@ -0,0 +1,456 @@
+              }
+            }
+          },
+        },
+      );
+    },
+    roll(currentEvent: OntimeEvent | null, nextEvent: OntimeEvent | null) {
+      mutate((state) => {
+        state.timer.clock = clock.timeNow();
+
+        if (currentEvent) {
+          // there is something running, load
+          state.timer.secondaryTimer = null;
+          state.timer.secondaryTarget = null;
+
+          // account for event that finishes the day after
+          const endTime =
+            currentEvent.timeEnd < currentEvent.timeStart ? currentEvent.timeEnd + dayInMs : currentEvent.timeEnd;
+
+          // when we load a timer in roll, we do the same things as before
+          // but also pre-populate some data as to the running state
+          this.load(currentEvent, {
+            startedAt: currentEvent.timeStart,
+            expectedFinish: currentEvent.timeEnd,
+            current: endTime - state.timer.clock,
+          });
+        } else if (nextEvent) {
+          // account for day after
+          const nextStart =
+            nextEvent.timeStart < state.timer.clock ? nextEvent.timeStart + dayInMs : nextEvent.timeStart;
+          // nothing now, but something coming up
+          state.timer.secondaryTimer = nextStart - state.timer.clock;
+          state.timer.secondaryTarget = nextStart;
+        }
+
+        state.playback = Playback.Roll;
+      });
+    },
+  },
+};
+
+/**
+ * This function is the only way to write to the state. You should mock this
+ * function on the unit tests to assert state transitions and side effects.
+ */
+function mutate<R>(
+  /**
+   * A pure function that receives the current state and modifies it in place.
+   * The function can return a value that will be passed to the side effects, and
+   * will be returned to the caller of the mutation.
+   */
+  fn: (state: DeepWritable<TState>) => R,
+  opts?: {
+    /**
+     * Side effects are functions that execute code related to the change in state.
+     *
+     * @param result The result of the mutation function `fn`
+     * @param newState The new state after the mutation
+     * @param prevState The state before the mutation
+     */
+    sideEffect?: (result: R, newState: TState, prevState: TState) => void;
+  },
+) {
+  const prevState = state;
+
+  const result = fn(state as DeepWritable<TState>);
+
+  const newState = state;
+
+  opts?.sideEffect?.(result, newState, prevState);
+
+  try {
+    eventStore.batchSet({
+      playback: newState.playback,
+      timer: newState.timer,
+    });
+  } catch (error) {
+    const shouldSwallow =
+      error instanceof Error && error.message === "Cannot access 'eventStore' before initialization";
+
+    if (!shouldSwallow) {
+      throw error;
+    }
+  }
+
+  restoreService.save({
+    playback: state.playback,
+    selectedEventId: state.timer.selectedEventId,
+    startedAt: state.timer.startedAt,
+    addedTime: state.timer.addedTime,
+    pausedAt: state.timer.pausedAt,
+  });
+
```

Comment:
I have separed this into an utility function [here](https://github.com/cpvalente/ontime/pull/692). Could we devise a more elegant way to write utilities that the state mutations use?